### PR TITLE
Write the output of make-[tables|generators] to a file

### PR DESCRIPTION
**Pull request by Iain-S**
We currently write the output of the `make-` commands to stdout but this doesn't work so well now that we also write other files (e.g. vocab `.csv` files during make-generators and, potentially, `stats.yaml`).

This is a big single commit where we:

1. Write the ORM and SSG files directly to disk (rather than stdout)
1. Have defaults for the ORM and SSG file names, which can be overridden with `--ssg-file=myfilename.py`, etc.
1. `sys.exit(1)` if an ORM, SSG or vocab file we are trying to write already exists (up for debate)
1. Import files from the local directory (before, I think we were importing them relative to `main.py` or `make.py`, which lead to some funny behaviour. We now require the file to be in the current working dir, which necessitated several test changes.
1. Rename `tests/tmp` to `tests/workspace` and run our functional tests there.
1. Swap `with patch("") as mock_something` context managers for `@patch("")` test function decorators for better readability (and quicker to create/delete patch calls this way).

For the reviewer:

1. Does this work end-to-end for you?

### CODE REVIEW

File: tests/test_functional.py [hunk 1]
```diff
@@ -5,56 +5,162 @@
...
     def setUp(self) -> None:
         """Pre-test setup."""
-        self.orm_file_path.unlink(missing_ok=True)
-        self.ssg_file_path.unlink(missing_ok=True)
 
+        # Create a blank destination database
         run_psql("dst.dump")
 
-    def test_workflow(self) -> None:
+        os.chdir("tests/workspace")
```
Comment (mhauru) [inline tests/test_functional.py:42]: I wanted to use a `tmp_path` pytest fixture for this, to use a temporary folder that shouldn't conflict with anything existing, but pytest fixtures don't play nicely with `unittest.TestCase` subclasses. (https://docs.pytest.org/en/6.2.x/unittest.html)

File: tests/test_main.py [hunk 2]
```diff
@@ -24,173 +22,158 @@ def assertSuccess(self, result: Result) -> None:
...
+        """Test the make-generators sub-command."""
+        with patch("sqlsynthgen.main.make_generators_from_tables") as mock_make, patch(
+            "sqlsynthgen.main.Path"
+        ) as mock_path, patch("sqlsynthgen.main.import_file") as mock_import, patch(
+            "sqlsynthgen.main.read_yaml_file"
+        ):
+            mock_path.return_value.exists.return_value = False
+            mock_make.return_value = "some text"
+
+            # conf_path = "tests/examples/generator_conf.yaml"
```
Comment (mhauru) [inline tests/test_main.py:51]: Can be removed?

File: tests/test_main.py [hunk 3]
```diff
@@ -24,173 +22,158 @@ def assertSuccess(self, result: Result) -> None:
...
+            # with open(conf_path, "r", encoding="utf8") as f:
+            #     config = yaml.safe_load(f)
 
             result = runner.invoke(
                 app,
                 [
-                    "make-tables",
+                    "make-generators",
+                    # conf_path,
```
Comment (mhauru) [inline tests/test_main.py:59]: Can be removed?

**Review by mhauru — APPROVED** [review]
One minor request and one general comment, good otherwise

File: tests/test_functional.py (hunk 1 above)
Comment (Iain-S) [reply tests/test_functional.py:42]: Yeah, I think there must be a nicer way to do this than `chdir()` in the test functions / in the setUp. I could imagine the current way might cause some interdependence between tests if one fails in the setup before the first `chdir()`.

**Pull request merged**